clash-lib/tests/Test/Clash/Rewrite.hs

```diff
@@ -0,0 +1,301 @@
+import Data.Default
+import Language.Haskell.Exts.Syntax
+import Language.Haskell.Exts.Parser (parseExp, fromParseResult)
+import System.IO.Unsafe (unsafePerformIO)
+import Text.Read (readMaybe)
+import GHC.Stack (HasCallStack)
+
+import qualified Language.Haskell.TH.Syntax as TH
+import qualified Language.Haskell.TH.Quote as TH
+
+import qualified Data.List as List
+import qualified Data.HashMap.Strict as HashMap
+import qualified Data.Map as Map
+import qualified Data.Set as Set
+import qualified Data.IntMap as IntMap
+import qualified Data.Text as Text
+
+type TypeMap = HashMap.HashMap Int C.Type
+
+lookupTM :: Int -> TypeMap -> C.Type
+lookupTM u tm = case HashMap.lookup u tm of
+  Just t -> t
+  Nothing ->
+    error [I.i|
+      Tried to lookup unique '#{u}' in typemap, but couldn't find it. This
+      usually means you forgot to (explicitely) declare a variable's type.
+    |]
+
+instance Default RewriteEnv where
+  def = RewriteEnv
+    { _dbgLevel=DebugSilent
+    , _dbgTransformations=Set.empty
+    , _dbgTransformationsFrom=0
+    , _dbgTransformationsLimit=maxBound
+    , _aggressiveXOpt=False
+    , _typeTranslator=error "_typeTranslator: NYI"
+    , _tcCache=emptyUniqMap
+    , _tupleTcCache=IntMap.empty
+    , _evaluator=error "_evaluator: NYI"
+    , _topEntities=emptyVarSet
+    , _customReprs=buildCustomReprs []
+    }
+
+instance Default extra => Default (RewriteState extra) where
+  def = RewriteState
+    { _transformCounter=0
+    , _bindings=emptyVarEnv
+    , _uniqSupply=unsafePerformIO newSupply
+    , _curFun=error "_curFun: NYI"
+    , _nameCounter=2
+    , _globalHeap=error "_globalHeap: NYI"
+    , _extra=def
+    }
+
+instance Default NormalizeState where
+  def = NormalizeState
+    { _normalized=emptyVarEnv
+    , _specialisationCache=Map.empty
+    , _specialisationHistory=emptyVarEnv
+    , _specialisationLimit=20
+    , _inlineHistory=emptyVarEnv
+    , _inlineLimit=20
+    , _inlineFunctionLimit=15
+    , _inlineConstantLimit=0
+    , _primitives=HashMap.empty
+    , _primitiveArgs=Map.empty
+    , _recursiveComponents=emptyVarEnv
+    , _newInlineStrategy=True
+    , _normalizeUltra=False
+    , _inlineWFCacheLimit=10
+    }
+
+instance Default InScopeSet where
+  def = emptyInScopeSet
+
+-- | Run a single transformation given a certain context
+runSingleTransformation
+  :: RewriteEnv
+  -- ^ Rewrite environment
+  -> RewriteState extra
+  -- ^ Rewrite state
+  -> InScopeSet
+  -- ^ Variables in scope in transformation
+  -> Rewrite extra
+  -- ^ Transformation to perform
+  -> C.Term
+  -- ^ Term to transform
+  -> C.Term
+runSingleTransformation rwEnv rwState is trans term = t
+ where (t, _, _) = runR (runRewrite "" is trans term) rwEnv rwState
+
+-- | Run a single transformation with an empty environment and empty
+-- InScopeSet. See Default instances ^ to inspect the precise definition of
+-- "empty".
+runSingleTransformationDef :: Default extra => Rewrite extra -> C.Term -> C.Term
```

Comment:
This functions should come with a BIG WARNING telling the developer not to use it for transformations that:
1. Evaluate to terms to WHNF, e.g. `caseCon`
2. Make decisions based on the "representability" of types, e.g. `inlineNonRep`
Since the evaluator and type translator are bottoming for the Default instance.

Reply:
They'd error out almost immediately, so I'm not too worried. I'll add a warning anyway though!